packages/ckeditor5-image/src/imageutils.js

```diff
@@ -0,0 +1,345 @@
+	}
+
+	/**
+	 * Checks if a given view element is an image widget.
+	 *
+	 * @protected
+	 * @param {module:engine/view/element~Element} viewElement
+	 * @returns {Boolean}
+	 */
+	isImageWidget( viewElement ) {
+		return !!viewElement.getCustomProperty( 'image' ) && isWidget( viewElement );
+	}
+
+	/**
+	 * Checks if the provided model element is an `image`.
+	 *
+	 * @protected
+	 * @param {module:engine/model/element~Element} modelElement
+	 * @returns {Boolean}
+	 */
+	isBlockImage( modelElement ) {
+		return !!modelElement && modelElement.is( 'element', 'image' );
+	}
+
+	/**
+	 * Checks if the provided model element is an `imageInline`.
+	 *
+	 * @protected
+	 * @param {module:engine/model/element~Element} modelElement
+	 * @returns {Boolean}
+	 */
+	isInlineImage( modelElement ) {
+		return !!modelElement && modelElement.is( 'element', 'imageInline' );
+	}
+
+	/**
+	 * Get view `<img>` element from the view widget (`<figure>`).
+	 *
+	 * Assuming that image is always a first child of a widget (ie. `figureView.getChild( 0 )`) is unsafe as other features might
+	 * inject their own elements to the widget.
+	 *
+	 * The `<img>` can be wrapped to other elements, e.g. `<a>`. Nested check required.
+	 *
+	 * @protected
+	 * @param {module:engine/view/element~Element} figureView
+	 * @returns {module:engine/view/element~Element}
+	 */
+	getViewImageFromWidget( figureView ) {
+		if ( this.isInlineImageView( figureView ) ) {
+			return figureView;
+		}
+
+		const figureChildren = [];
+
+		for ( const figureChild of figureView.getChildren() ) {
+			figureChildren.push( figureChild );
+
+			if ( figureChild.is( 'element' ) ) {
+				figureChildren.push( ...figureChild.getChildren() );
+			}
+		}
+
+		return figureChildren.find( this.isInlineImageView );
+	}
+}
+
+// Checks if image is allowed by schema in optimal insertion parent.
+//
+// @private
+// @param {module:core/editor/editor~Editor} editor
+// @param {module:engine/model/selection~Selection} selection
+// @returns {Boolean}
+function isImageAllowedInParent( editor, selection ) {
+	const imageType = determineImageTypeForInsertion( editor, selection );
+
+	if ( imageType == 'image' ) {
+		const parent = getInsertImageParent( selection, editor.model );
+
+		if ( editor.model.schema.checkChild( parent, 'image' ) ) {
+			return true;
+		}
+	} else if ( editor.model.schema.checkChild( selection.focus, 'imageInline' ) ) {
+		return true;
+	}
+
+	return false;
+}
+
+// Checks if selection is not placed inside an image (e.g. its caption).
+//
+// @private
+// @param {module:engine/model/selection~Selectable} selection
+// @returns {Boolean}
+function isNotInsideImage( selection ) {
+	return [ ...selection.focus.getAncestors() ].every( ancestor => !ancestor.is( 'element', 'image' ) );
+}
+
+// Returns a node that will be used to insert image with `model.insertContent`.
+//
+// @private
+// @param {module:engine/model/selection~Selectable} selection
```

Comment:
```suggestion
// @param {module:engine/model/selection~Selection} selection
```

This could be also `DocumentSelection` but since this is internal helper we don't need to focus on that (same applies to other functions - I just skipped comments there)